src/zip_longest.rs

```diff
@@ -52,6 +52,32 @@ where
     fn size_hint(&self) -> (usize, Option<usize>) {
         size_hint::max(self.a.size_hint(), self.b.size_hint())
     }
+
+    #[inline]
+    fn fold<B, F>(self, mut acc: B, mut f: F) -> B
+    where
+        Self: Sized,
+        F: FnMut(B, Self::Item) -> B,
+    {
+        let ZipLongest { mut a, mut b } = self;
+
+        loop {
+            match (a.next(), b.next()) {
+                (Some(x), Some(y)) => acc = f(acc, EitherOrBoth::Both(x, y)),
+                (Some(x), None) => {
+                    acc = f(acc, EitherOrBoth::Left(x));
+                    // b is exhausted, so we can drain a.
+                    return a.fold(acc, |acc, x| f(acc, EitherOrBoth::Left(x)));
+                }
+                (None, Some(y)) => {
+                    acc = f(acc, EitherOrBoth::Right(y));
+                    // a is exhausted, so we can drain b.
+                    return b.fold(acc, |acc, y| f(acc, EitherOrBoth::Right(y)));
+                }
+                (None, None) => return acc, // Both iterators are exhausted.
+            }
+        }
+    }
```

Comment:
When specializing `fold`, try to avoid `for` loops and calls to `next()` on underlying iterators. If those underlying iterators also specialize `fold`, then `for` looping and calling `next()` bypasses their optimizations.

It's possible to implement this method in terms of fold:
```suggestion
    #[inline]
    fn fold<B, F>(self, mut init: B, mut f: F) -> B
    where
        Self: Sized,
        F: FnMut(B, Self::Item) -> B,
    {
        let ZipLongest { a, mut b } = self;

        init = a.fold(init, |init, a| match b.next() {
            Some(b) => f(init, EitherOrBoth::Both(a, b)),
            None => f(init, EitherOrBoth::Left(a)),
        });

        b.fold(init, |init, b| f(init, EitherOrBoth::Right(b)))
    }
```

Reply:
Thanks @jswrenn for your feedback. I tried to run the benchmark against your recommendation but the result is regressing

```
time:   [594.12 ns 594.85 ns 595.65 ns]
change: [+20.625% +21.035% +21.434%] (p = 0.00 < 0.05)
Performance has regressed.
```

Reply:
I was gonna suggest to try to rely a bit on `zip` from the std with `init = a.by_ref().zip(b.ref()).fold(init, &mut f); ...` but `Zip` do not have a `fold` specialization yet. It would allow to discard the loop (and the `Some, Some` case).

What about `f(init, match b.next() {...})` (in the first fold)? (I guess it's the same?)

Reply:
Yes unfortunately. I tried : 

```
{
        let ZipLongest { a, mut b } = self;
        init = a.fold(init, |init, a| f(init, match b.next() {
            Some(b) => EitherOrBoth::Both(a, b),
            None => EitherOrBoth::Left(a),
        }));
        b.fold(init, |init, b| f(init, EitherOrBoth::Right(b)))
    }
```
and the result is same: `time:   [593.41 ns 594.82 ns 596.39 ns]`

Reply:
In #765 I considered folding a range `k+1..=n` but did not because it was slower, which surprised me. It seems weird to even suggest but maybe it's because of `Range*::fold`?!
Since `fold`s in the benchmark rely on the same thing here, maybe you can try with different iterators.
Like `let v1 = black_box((0..***).collect_vec()); v2...` (before `bench_function`) and then `v1.iter().zip_longest(v2.iter())`.

Reply:
The performance difference of  a `fold` specialization is going to depend on the relative performance of calling `fold` vs `next` on the underlying iterators. If the underlying iterator involves adapters like `chain` or `intersperse`, then using `fold` will be much faster.

The reason for this is that the `next()` calls of `chain` and `intersperse` have a branch on every call; e.g.:
https://github.com/rust-lang/rust/blob/a66e334bc5c5395907dddfc6211cbb110c44576d/library/core/src/iter/adapters/chain.rs#L49C14-L52

...which is avoided in their `fold` specialization:
https://github.com/rust-lang/rust/blob/a66e334bc5c5395907dddfc6211cbb110c44576d/library/core/src/iter/adapters/chain.rs#L85-L96

As a rule, I think we should optimize for the assumption that calling `fold` on underlying iterators will be faster than calling `next`. (After all, the end-user is calling `fold` because *they* assume, for some reason, that internal iteration will be faster than external iteration.)

Reply:
@jswrenn Thanks a lot! Yes after looking at the 2 files you shared, I can understand why for loop should always be avoided (and use the underlying fold api which has been optimized). That's really inspiring to me. I've incorporated your suggestions just now. Thanks a lot!

@Philippe-Cholet Yes the issue is on `Range*::fold`. I adopted your suggestions and rerun the benchmark. Using the default `fold` (which calls `next`) the time is `time:   [2.0522 µs 2.0558 µs 2.0599 µs]`, and using the new `fold` logic, I can see a huge improvement `time:   [414.34 ns 415.44 ns 416.75 ns]`

Thanks both!

Reply:
Diving... Apparently, Range (`a..b`) does not specialize fold (so maybe that's why there was bad perf), but RangeInclusive (`a..=b`) [does](https://doc.rust-lang.org/stable/src/core/iter/range.rs.html#1160) relying on `try_fold` (implementation seems tricky to me). `std::vec::IntoIter` does not either, but `std::slice::Iter` which I redirected you to (without much insight) does.
So I guess we should continue with `some_vec.iter().fold(...)` for benchmarks.

Reply:
Indeed. In future PR that optimizes `fold`, we should stick with `some_vec.iter().fold(...)` for benchmarking

Reply:
@jswrenn Do you have any further feedback for my PR ? If no I think we can merge this.

Reply:
Usual link for how things should prefer forwarding internal iteration to internal iteration where possible: <https://medium.com/@veedrac/rust-is-slow-and-i-am-the-cure-32facc0fdcb>.

(So this is my +1 to forwarding `fold` to `fold` where possible, and when we finally make it possible to implement `try_fold`, having that forward to `try_fold`.)